File: features/step_definitions/camera.rb
# Step definitions for camera operations

Given("hsize ← {int}") do |value|
  @hsize = value
end

Given("vsize ← {int}") do |value|
  @vsize = value
end

Given(/field_of_view ← π\/(\d+)/) do |divisor|
  @field_of_view = Math::PI / divisor.to_f
end

When("c ← camera\\(hsize, vsize, field_of_view)") do
  @c = Rayz::Camera.new(hsize: @hsize, vsize: @vsize, field_of_view: @field_of_view)
end

Then("c.hsize = {int}") do |value|
  assert_equal(@c.hsize, value)
end

Then("c.vsize = {int}") do |value|
  assert_equal(@c.vsize, value)
end

Then(/c\.field_of_view = π\/(\d+)/) do |divisor|
  expected = Math::PI / divisor.to_f
  assert_equal(@c.field_of_view, expected)
end

Then("c.transform = identity_matrix") do
  assert_equal(@c.transform, Matrix.identity(4))
end

Given(/c ← camera\((\d+), (\d+), π\/(\d+)\)/) do |hsize, vsize, divisor|
  @c = Rayz::Camera.new(hsize: hsize.to_i, vsize: vsize.to_i, field_of_view: Math::PI / divisor.to_f)
end

Then("c.pixel_size = {float}") do |value|
  assert_in_delta(@c.pixel_size, value, Rayz::Util::EPSILON)
end

When("r ← ray_for_pixel\\(c, {int}, {int})") do |x, y|
  @r = @c.ray_for_pixel(x, y)
end

When(/c\.transform ← rotation_y\(π\/(\d+)\) \* translation\(([^,]+),\s*([^,]+),\s*([^)]+)\)/) do |pi_divisor, tx, ty, tz|
  @c.transform = Rayz::Transformations.rotation_y(radians: Math::PI / pi_divisor.to_f) * Rayz::Transformations.translation(x: tx.to_f, y: ty.to_f, z: tz.to_f)
end

Then(/r\.direction = vector\(√(\d+)\/(\d+), (\d+(?:\.\d+)?), -√(\d+)\/(\d+)\)/) do |sqrt_num_x, sqrt_den_x, y, sqrt_num_z, sqrt_den_z|
  expected = Rayz::Vector.new(
    x: Math.sqrt(sqrt_num_x.to_f) / sqrt_den_x.to_f,
    y: y.to_f,
    z: -Math.sqrt(sqrt_num_z.to_f) / sqrt_den_z.to_f
  )
  assert_equal(@r.direction, expected)
end

Given("from ← point\\({float}, {float}, {float})") do |x, y, z|
  @from = Rayz::Point.new(x: x, y: y, z: z)
end

Given("to ← point\\({float}, {float}, {float})") do |x, y, z|
  @to = Rayz::Point.new(x: x, y: y, z: z)
end

Given("up ← vector\\({float}, {float}, {float})") do |x, y, z|
  @up = Rayz::Vector.new(x: x, y: y, z: z)
end

Given("c.transform ← view_transform\\(from, to, up)") do
  @c.transform = Rayz::Transformations.view_transform(from: @from, to: @to, up: @up)
end

When("image ← render\\(c, w)") do
  @image = @c.render(@w)
end

Then("pixel_at\\(image, {int}, {int}) = color\\({float}, {float}, {float})") do |x, y, r, g, b|
  expected = Rayz::Color.new(red: r, green: g, blue: b)
  actual = @image.pixel_at(col: x, row: y)
  assert_equal(actual, expected)
end

File: python/features/steps/camera.py
import pytest
from behave import given, then, use_step_matcher, when

from rayz.camera import Camera
from rayz.color import Color
from rayz.math_parser import parse_math
from rayz.matrix import Matrix
from rayz.transformations import rotation_y, translation, view_transform

use_step_matcher("re")

_V = r"([A-Za-z][A-Za-z0-9_]*)"
_A = r"([^\s,)]+)"

IDENTITY_MATRIX = Matrix.identity(4)


def _val(context, s):
    """Return numeric value: look up as context attribute if it's a plain name."""
    if hasattr(context, s):
        return getattr(context, s)
    return parse_math(s)


def _make_camera(context, var, h, v, fov):
    setattr(context, var, Camera(int(_val(context, h)), int(_val(context, v)), _val(context, fov)))


# ---------------------------------------------------------------------------
# Given
# ---------------------------------------------------------------------------


@given(rf"hsize ← {_A}")
def step_given_hsize(context, val):
    context.hsize = int(parse_math(val))


@given(rf"vsize ← {_A}")
def step_given_vsize(context, val):
    context.vsize = int(parse_math(val))


@given(rf"field_of_view ← {_A}")
def step_given_fov(context, val):
    context.field_of_view = parse_math(val)


@given(rf"{_V} ← camera\({_A},\s*{_A},\s*{_A}\)")
def step_given_camera(context, var, h, v, fov):
    _make_camera(context, var, h, v, fov)


@given(rf"{_V}\.transform ← view_transform\({_V},\s*{_V},\s*{_V}\)")
def step_given_camera_view_transform(context, cam_var, from_var, to_var, up_var):
    getattr(context, cam_var).transform = view_transform(
        getattr(context, from_var),
        getattr(context, to_var),
        getattr(context, up_var),
    )


# ---------------------------------------------------------------------------
# When
# ---------------------------------------------------------------------------


@when(rf"{_V} ← camera\({_A},\s*{_A},\s*{_A}\)")
def step_when_camera(context, var, h, v, fov):
    _make_camera(context, var, h, v, fov)


@when(rf"{_V}\.transform ← rotation_y\({_A}\) \* translation\({_A},\s*{_A},\s*{_A}\)")
def step_when_camera_transform(context, var, ry_val, tx, ty, tz):
    t = rotation_y(parse_math(ry_val)) * translation(parse_math(tx), parse_math(ty), parse_math(tz))
    getattr(context, var).transform = t


@when(rf"r ← ray_for_pixel\({_V},\s*{_A},\s*{_A}\)")
def step_when_ray_for_pixel(context, cam_var, px, py):
    context.r = getattr(context, cam_var).ray_for_pixel(int(parse_math(px)), int(parse_math(py)))


@when(rf"image ← render\({_V},\s*{_V}\)")
def step_when_render(context, cam_var, world_var):
    context.image = getattr(context, cam_var).render(getattr(context, world_var))


# ---------------------------------------------------------------------------
# Then
# ---------------------------------------------------------------------------


@then(rf"{_V}\.hsize = {_A}")
def step_then_hsize(context, var, val):
    assert getattr(context, var).hsize == int(parse_math(val))


@then(rf"{_V}\.vsize = {_A}")
def step_then_vsize(context, var, val):
    assert getattr(context, var).vsize == int(parse_math(val))


@then(rf"{_V}\.field_of_view = {_A}")
def step_then_fov(context, var, val):
    assert getattr(context, var).field_of_view == pytest.approx(parse_math(val), abs=1e-5)


@then(rf"{_V}\.transform = identity_matrix")
def step_then_camera_transform_identity(context, var):
    assert getattr(context, var).transform == IDENTITY_MATRIX


@then(rf"{_V}\.pixel_size = {_A}")
def step_then_pixel_size(context, var, val):
    assert getattr(context, var).pixel_size == pytest.approx(parse_math(val), abs=1e-4)


@then(rf"pixel_at\(image,\s*{_A},\s*{_A}\) = color\({_A},\s*{_A},\s*{_A}\)")
def step_then_pixel_at(context, px, py, cr, cg, cb):
    pixel = context.image.pixel_at(col=int(parse_math(px)), row=int(parse_math(py)))
    expected = Color(parse_math(cr), parse_math(cg), parse_math(cb))
    assert pixel == expected, f"{pixel!r} != {expected!r}"
